### ISSUE TITLE
Flaky tests in ClientServerSpec

### ISSUE THREAD

**Issue by 2beaucoup** [akka/akka#18005]
The <code>log materialization errors in `bindAndHandle`</code> tests fail from time to time with failures like this:

```
java.lang.AssertionError: assertion failed: expected no StreamSupervisor children, but got [Actor[akka://ClientServerSpec/user/$a/flow-8-15-actorPublisherSource-actorPublisherSource-actorPublisherSource#748355490], Actor[akka://ClientServerSpec/user/$a/flow-8-25-errorLogger#1211291213]]
	at scala.Predef$.assert(Predef.scala:179)
	at akka.stream.testkit.Utils$$anonfun$assertAllStagesStopped$1$$anonfun$apply$mcV$sp$1.apply$mcV$sp(Utils.scala:31)
	at akka.stream.testkit.Utils$$anonfun$assertAllStagesStopped$1$$anonfun$apply$mcV$sp$1.apply(Utils.scala:28)
	at akka.stream.testkit.Utils$$anonfun$assertAllStagesStopped$1$$anonfun$apply$mcV$sp$1.apply(Utils.scala:28)
	at akka.testkit.TestKitBase$class.poll$2(TestKit.scala:254)
	at akka.testkit.TestKitBase$class.awaitAssert(TestKit.scala:265)
	at akka.testkit.TestKit.awaitAssert(TestKit.scala:718)
	at akka.stream.testkit.Utils$$anonfun$assertAllStagesStopped$1.apply$mcV$sp(Utils.scala:28)
	at akka.stream.testkit.Utils$$anonfun$assertAllStagesStopped$1.apply(Utils.scala:28)
	at akka.stream.testkit.Utils$$anonfun$assertAllStagesStopped$1.apply(Utils.scala:28)
	at akka.testkit.TestKitBase$class.within(TestKit.scala:296)
	at akka.testkit.TestKit.within(TestKit.scala:718)
	at akka.testkit.TestKitBase$class.within(TestKit.scala:310)
	at akka.testkit.TestKit.within(TestKit.scala:718)
	at akka.stream.testkit.Utils$.assertAllStagesStopped(Utils.scala:27)
	at akka.http.scaladsl.ClientServerSpec$$anonfun$1$$anonfun$apply$mcV$sp$5$$anonfun$apply$mcV$sp$9.apply$mcV$sp(ClientServerSpec.scala:179)
	at akka.http.scaladsl.ClientServerSpec$$anonfun$1$$anonfun$apply$mcV$sp$5$$anonfun$apply$mcV$sp$9.apply(ClientServerSpec.scala:179)
	at akka.http.scaladsl.ClientServerSpec$$anonfun$1$$anonfun$apply$mcV$sp$5$$anonfun$apply$mcV$sp$9.apply(ClientServerSpec.scala:179)
```

I'm currenlty on Wndows and played with the timeouts but didn't get them to pass.

**Comment by jrudolph** [comment]
Did you change the timeouts in `assertAllStagesStopped`?

**Comment by 2beaucoup** [comment]
@jrudolph Doing that didn't help either.

**Comment by jrudolph** [comment]
This could then be a valid potential memory leak which is only triggered under the right circumstances.

**Comment by drewhk** [comment]
No, this is just the fact that materializer is attempting to clean up a partially constructed graph in panic mode, which is best effort. There is no really reliable way of doing this unfortunately. For actors the cleanup can be a bit more robust by adding a supervisor actor for each materialization and shutting down that one, but this does not help with non-actor backed RS entities. This test is flaky, so probably we need to remove the assertAllStagesStopped marker.

**Comment by jrudolph** [comment]
@drewhk to what did you respond with "No"? I.e. which of the statements hold:

 * `assertAllStagesStopped` is reliably reporting leaked actors
 * even in "panic mode" all actors are cleaned up properly
 * `ClientServerSpec` does never leak any actor instances

So, in summary, is there a potential leak or is this only a testing issue?

**Comment by drewhk** [comment]
@jrudolph my answers

* assertAllStagesStopped is reliably reporting leaked actors

Yes, it is reliably reporting them.

* even in "panic mode" all actors are cleaned up properly

No, panic mode is best effort, resources might leak. One way to solve this is to add an additional supervision layer (and +1 actor) per materialization, then in panic shut down that actor. It wont prevent leakage of non-actor based 3rd party RS instances though.

* ClientServerSpec does never leak any actor instances

No, it does not since the ActorSystem is shut down, and assertAllStagesStopped always clears previous leftovers (so failing tests don't infect other tests).

**Comment by jrudolph** [comment]
Ok, what does that mean for real world cases? It has been an long time we talked about this so I don't remember clearly. In which cases is panic mode invoked and does it happen in the wild? How do users avoid memory leaks?

**Comment by drewhk** [comment]
Currently, if you fail the materialization somehow (like a stage factory function is throwing) then the graph is in a partially constructed state, so in "panic" mode the materializer just sends onError() and cancel() to the Publishers and Subscribers it knows about. This is a band-aid, and solves most of the issues, but might leak actors. The problem is that the onError paths are not established at this point so there is no easy Spec compliant way to shut down things.

A better solution will be that instead of attaching all actors corresponding to a given materialization to the supervisor actor of the Materializer, we attach first a sub-supervisor that will in turn contain/parent all the materialized actors for that materialization. In panic, we just kill that actor and that ensures we don't leak any actors. OTOH it will not ensure that the non-actor backed RS entites will be cleaned up properly.

**Comment by drewhk** [comment]
tl;dr;
 - originally, there was a definite leak if materialization fail
 - panic mode was added, which mitigates this leak, eliminating it most of the time (not 100% though)
 - additional hierarchy level will 100% solve the actor leak case, only the non-actor based might be leaking, but we can mitigate that as before

**Comment by patriknw** [comment]
closing because it has not reoccurred in a long time
